lib/Sema/MiscDiagnostics.cpp

```diff
@@ -1282,25 +1282,19 @@ static void diagSyntacticUseRestrictions(const Expr *E, const DeclContext *DC,
 
     }
     
-    /// Return true if this is a 'nil' literal.  This looks
-    /// like this if the type is Optional<T>:
-    ///
-    ///   (dot_syntax_call_expr implicit type='Int?'
-    ///     (declref_expr implicit decl=Optional.none)
-    ///     (type_expr type=Int?))
-    ///
-    /// Or like this if it is any other ExpressibleByNilLiteral type:
-    ///
-    ///   (nil_literal_expr)
-    ///
-    bool isTypeCheckedOptionalNil(Expr *E) {
+    /// Returns true if this is a 'nil' literal
+    bool isNilLiteral(Expr *E) {
       if (dyn_cast<NilLiteralExpr>(E)) return true;
```

Comment:
Not necessary to have a separate function that's a synonym for `dyn_cast<NilLiteralExpr>` if you're refactoring this out.

That said, seems like besides the diagnostic change the substantive modification here is deletion of `|| !CE->isImplicit())`, and the rest of the code can stand as-is? Would be good to have the test that @LucianoPAlmeida mentioned as well as any others that exercise the difference that happens when the `isImplicit()` check is removed. Any code organization changes can be a separate PR to leave the substantive changes here clear. As to the separate diagnostic, I think it'd be fine to have just one that says `'nil' or 'Optional.none'`.

Reply:
@xwu, I combined the diagnostics as you suggested. As for removing the `|| !CE->isImplicit())` condition -- that's the specific change that causes this new warning to be emitted.

The docs for the `isTypeCheckedOptionalNil` method say the expression for an `Optional.none` case looks like this:

```
(dot_syntax_call_expr implicit type='Int?'     // <- implicit
  (declref_expr implicit decl=Optional.none)
  (type_expr type=Int?))
```

That comment is from 6 years ago so it may be outdated -- when I dump the expression I instead see:

```
(dot_syntax_call_expr type='String?'  // <- not implicit
  (declref_expr type='(Optional<String>.Type) -> Optional<String>' decl=Swift.(file).Optional.none function_ref=unapplied)
  (argument_list implicit
    (argument
      (type_expr implicit type='String?.Type' typerepr='String?'))))
```

Removing the `|| !CE->isImplicit())` condition is necessary now that the `DotSyntaxCallExpr` is not implicit. This change seems safe to me considering it doesn't cause any new unexpected warnings. Can you think of any additional cases that need to be tested?

Reply:
@calda Thanks, as I'm not familiar with why `isImplicit` was there in the first place, it's hard for me to know what to suggest here.

My concern wasn't whether deleting that expression produces false negatives (i.e., fails to warn when it should) but whether deleting that will incorrectly now produce any false positives (i.e., warns about comparison to `Optional.none` when that's not what's happening). That really depends on knowing what `isImplicit` was put in to reject way back when—it seems like it served _some_ purpose because otherwise it'd not have been tested for, and now that it's gone one wonders what isn't being rejected anymore.